src/lightning/fabric/wrappers.py

```diff
@@ -168,20 +169,17 @@ def __len__(self) -> int:
         return len(self._dataloader)
 
     def __iter__(self) -> Union[Iterator[Any], Generator[Any, None, None]]:
-        if hasattr(self._dataloader.sampler, "set_epoch"):
-            # Without setting the epoch, the distributed sampler would return the same indices every time, even when
-            # shuffling is enabled. In PyTorch, the user would normally have to call `.set_epoch()` on the sampler.
-            # In Lite, we take care of this boilerplate code.
-            self._dataloader.sampler.set_epoch(self._num_iter_calls)
+        # Without setting the epoch, the distributed sampler would return the same indices every time, even when
+        # shuffling is enabled. In PyTorch, the user would normally have to call `.set_epoch()` on the sampler.
+        # In Fabric, we take care of this boilerplate code.
+        _set_sampler_epoch(self._dataloader, self._num_iter_calls)
         self._num_iter_calls += 1
 
-        iterator = iter(self._dataloader)
         if self._device is None:
-            yield from iterator
-            return
-
-        for item in iterator:
```

Comment:
There was a small inefficiency here: for loop internally calls `iter()`, so this would do it twice